src/webapi/gamepad.rs

```diff
@@ -0,0 +1,217 @@
+    ///
+    /// [(JavaScript docs)](https://developer.mozilla.org/en-US/docs/Web/API/Gamepad/id)
+    // https://w3c.github.io/gamepad/#dom-gamepad-id
+    #[inline]
+    fn id(&self) -> String {
+        js!(
+            return @{self.as_ref()}.id;
+        ).try_into().unwrap()
+    }
+
+    /// An auto-incrementing integer to uniquely identify a connected Gamepad.
+    ///
+    /// [(JavaScript docs)](https://developer.mozilla.org/en-US/docs/Web/API/Gamepad/index)
+    // https://w3c.github.io/gamepad/#dom-gamepad-index
+    #[inline]
+    fn index(&self) -> i32 {
+        js!(
+            return @{self.as_ref()}.index;
+        ).try_into().unwrap()
+    }
+
+    /// Is this gamepad connected to the system, powered on, and usable?
+    ///
+    /// [(JavaScript docs)](https://developer.mozilla.org/en-US/docs/Web/API/Gamepad/connected)
+    // https://w3c.github.io/gamepad/#dom-gamepad-connected
+    #[inline]
+    fn connected(&self) -> bool {
+        js!(
+            return @{self.as_ref()}.connected;
+        ).try_into().unwrap()
+    }
+
+    /// Monotonically increasing time this gamepad was updated.
+    ///
+    /// [(JavaScript docs)](https://developer.mozilla.org/en-US/docs/Web/API/Gamepad/timestamp)
+    // https://w3c.github.io/gamepad/#dom-gamepad-timestamp
+    #[inline]
+    fn timestamp(&self) -> f64 {
+        js!(
+            return @{self.as_ref()}.timestamp;
+        ).try_into().unwrap()
+    }
+
+    /// The mapping in use for this device.
+    ///
+    /// [(JavaScript docs)](https://developer.mozilla.org/en-US/docs/Web/API/Gamepad/mapping)
+    // https://w3c.github.io/gamepad/#dom-gamepad-mapping
+    #[inline]
+    fn mapping(&self) -> GamepadMappingType {
+        js!(
+            return @{self.as_ref()}.mapping;
+        ).try_into().unwrap()
+    }
+
+    /// Array of values for all axes of the gamepad.
+    ///
+    /// [(JavaScript docs)](https://developer.mozilla.org/en-US/docs/Web/API/Gamepad/axes)
+    // https://w3c.github.io/gamepad/#dom-gamepad-axes
+    #[inline]
+    fn axes(&self) -> Vec<f64> {
+        js!(
+            return @{self.as_ref()}.axes;
+        ).try_into().unwrap()
+    }
+
+    /// Array of button states for all buttons of the gamepad.
+    ///
+    /// [(JavaScript docs)](https://developer.mozilla.org/en-US/docs/Web/API/Gamepad/buttons)
+    // https://w3c.github.io/gamepad/#dom-gamepad-buttons
+    #[inline]
+    fn buttons(&self) -> Vec<GamepadButton> {
+        js!(
+            return @{self.as_ref()}.buttons;
+        ).try_into().unwrap()
+    }
+}
+
+/// A reference to a JavaScript object which implements the [IGamepad](trait.IGamepad.html)
+/// interface.
+///
+/// [(JavaScript docs)](https://developer.mozilla.org/en-US/docs/Web/API/Gamepad)
+// https://w3c.github.io/gamepad/#gamepad-interface
+#[derive(Clone, Debug, Eq, PartialEq, ReferenceType)]
+#[reference(instance_of = "Gamepad")]
+pub struct Gamepad( Reference );
+
+impl IGamepad for Gamepad {}
```

Comment:
Hmm... since those are the only impementations we probably should get rid of the `IGamepadButton` and `IGamepad` interfaces and just define all of those methods in a normal `impl` block.

Reply:
sounds good. done!